Lib/fontTools/ttLib/ttGlyphSet.py

```diff
@@ -126,15 +153,59 @@ def draw(self, pen):
         how that works.
         """
         glyph, offset = self._getGlyphAndOffset()
+
+        if self.glyphSet.locationStack:
+            offset = 0  # Offset should only apply at top-level
+
+        if glyph.isVarComposite():
+            self._drawVarComposite(glyph, pen, False)
+            return
+
         glyph.draw(pen, self.glyphSet.glyfTable, offset)
 
     def drawPoints(self, pen):
         """Draw the glyph onto ``pen``. See fontTools.pens.pointPen for details
         how that works.
         """
         glyph, offset = self._getGlyphAndOffset()
+
+        if self.glyphSet.locationStack:
+            offset = 0  # Offset should only apply at top-level
+
+        if glyph.isVarComposite():
+            self._drawVarComposite(glyph, pen, True)
+            return
+
         glyph.drawPoints(pen, self.glyphSet.glyfTable, offset)
 
+    def _drawVarComposite(self, glyph, pen, isPointPen):
+
+        from fontTools.ttLib.tables._g_l_y_f import (
+            VarComponentFlags,
+            VAR_COMPONENT_TRANSFORM_MAPPING,
+        )
+
+        for comp in glyph.components:
+
+            with self.glyphSet.pushLocation(
+                comp.location, comp.flags & VarComponentFlags.RESET_UNSPECIFIED_AXES
+            ):
+                try:
+                    pen.addVarComponent(
+                        comp.glyphName, comp.transform, self.glyphSet.rawLocation
```

Comment:
Maybe I should directly pass `comp.location` here. That would be more truthful to the data. Yeah, I think that's more correct.

Reply:
Oh now I remember why I can't do that. If I do that then I also need to pass the `RESET` flag. So I kept it higher level. I'm not quite happy with this though. Happy to discuss. @justvanrossum

Reply:
how about we decouple this discussion about enhancing the pen protocol to make it "aware" of varcomponents to a separate follow-up PR? I think before this changes, this PR was relatively self-contained (mostly about glyf and the rest simply accompanying changes to accommodate the latter), and almost at merging stage; whereas changing the pen protocol could have wider effects, maybe we need to think about it more.

Reply:
I don't immediately understand what a "raw" location means.. Passing through the comp.location _and_ the reset flag seems better to me, if the intent is to delegate to the pen implementation

Reply:
The reset flag is mostly an optimization, we were hoping to keep the pen protocol stuff higher level so it will not have to worry about it. Especially since the compiler can figure out the best setting for it.

That said, you've got a point about fleshing out the pen stuff in a separate PR. (Although it could also be that we're almost there.)

Reply:
ok, I got it now. It's more like the "actual" or "effective" varcomponent location at that particular node in tree, not simply its nominal location (i.e. as encoded in the glyf table) that takes into account its reset flag as well as its parents, basically abstracting away the whole reset flag business

Reply:
Yes. That's the idea.